### ISSUE TITLE
Router.map callbacks hung off Router constructor

### ISSUE THREAD

**Issue by jasonmit** [emberjs/ember.js#13840]
When you take into account fastboot, this becomes unsafe if you call Router.map mulitple times within your application.  A particular use case is we registered routes dynamically depending on user state and we do this in an instance-initializer.  Where this becomes a problem is each time a request is made the callback to Router.map is hung off the constructor and therefore gets invoked on subsequent requests (likely leaking memory as well).

https://github.com/emberjs/ember.js/blob/master/packages/ember-routing/lib/system/router.js#L89-L96

Are the lines in question and so I was wondering if anyone would be opened to me moving these callbacks elsewhere, such as the application instance.  This will have larger ramifications, so would like to gather interest or feedback first.

**Comment by nathanhammond** [comment]
Yep, definitely an issue. This is made more frustrating by the many layers everything here has to go through in order to function the way Ember would like. If you can come up with a quick patch I would be happy to review it and we would likely land it. Depending upon the complexity/invasiveness the core team will decide what branch (canary, beta, release) to target. Probably a late addition to the 2.8.0-beta.X series of which the first one is planned for next week.

If it becomes too ridiculous the `RouterDSL` is on my list of things to rework after finishing route-recognizer.

**Comment by jasonmit** [comment]
@nathanhammond right now I don't have a good idea on how I could patch it while keeping the API intact.  I was thinking of perhaps exposing a new API for registering callbacks that are concatted with `constructor.dslCallbacks` during `_initRouterJs`.

Let me know if you have any ideas and feel free to ping me on Slack.

**Comment by nathanhammond** [comment]
Everything in that area _should_ be considered private API. I don't believe we need to maintain API consistency in addressing this bug. (Somebody else please chime in and correct me if I'm wrong.)

The only addon I know of which really abuses this portion of the router is one of mine, [ember-route-alias, and even I don't touch it](https://github.com/nathanhammond/ember-route-alias/blob/master/addon/initializers/route-alias.js).

**Comment by jasonmit** [comment]
I meant to say I don't know of a way of resolving this without implementing a new API for registering routes designed to be used at runtime instead of going through the constructor method.

**Comment by nathanhammond** [comment]
You could change the way we save off those references such that they're captured in a closure that is only accessible during that run? I _think_ that should be possible.

**Comment by jasonmit** [comment]
We spoke offline, the outcome was to spike two possible solutions and discuss.

**Comment by jasonmit** [comment]
/cc @wycats 

Based on our discussion today.  Feel free to add your input since you floated a few ideas, the core of the issue being the need for Router.map's callback to handle conditional logic and a similar concept of `needs`.

**Comment by jasonmit** [comment]
This is what I've ended up with to remove the need of calling `Router.map` for each app instance.

https://github.com/jasonmit/ember-routermap-inject-service

It's obviously touching private APIs, and I'd be happy to land something like this in the API if others feel it might be useful.

Another option might be exposing a service that enables registering routes at app runtime safely.  This would end up replacing `Router.map`

```
// router.js
export default Router.extend({
  session: inject.service().
  map() {
    let session = this.get('session');

    if (session.get('isAdmin')) {
      this.route('admin', { path: '/' });
    } else {
      this.route('home', { path: '/' });
    }

    this.route('about', { path: '/' + session.translate('urls.about') });
    this.route('contact', { path: '/' + session.translate('urls.contact') });
  }
});
```

**Comment by nathanhammond** [comment]
So what you've done is make sure that each invocation is identical, allowing Ember to continue to hang its callback functions off of `Router.constructor.dslCallbacks`. I feel like this is a good workaround, but doesn't solve the underlying problem.

I feel like we should go ahead and fix this at the fundamental level of making sure we clean up after ourselves instead of working around it. @jasonmit did you attempt to head down that path as well before going with the addition to the `RouterDSL` prototype?

**Comment by jasonmit** [comment]
@nathanhammond no it certainly doesn't fix the issue at the core, it does though eliminate having to invoke `Router.map` with the app instance in scope though.  I did chat with @wycats and what he said was basically the ability to `needs` voids surfacing this bug.

The way I would fix this at the core boils down to never hanging the callback off the constructor to access the callback function(s).  Instead, just resolve a module which exports a function - the `Router.map` callback.

**Comment by nathanhammond** [comment]
@jasonmit I'd like to see that!

**Comment by rwjblue** [comment]
FWIW, we use that same pattern for engines.  `ember-engines` uses `addon/routes.js` that just exports the callback.

**Comment by jasonmit** [comment]
@rwjblue noticed this while researching the bug. I'll be sure to piggyback off that work versus reinventing the wheel.

**Comment by pixelhandler** [comment]
@jasonmit @rwjblue @nathanhammond is this still an issue, perhaps we should close; what do you think?

**Comment by rwjblue** [comment]
We need to add an assertion when `Router.map` is called _after_ an `ApplicationInstance` instance is created.
